### ISSUE TITLE
Deep Links docs are not clear enough on how to read URL on cold start on Windows and Linux

### ISSUE THREAD

**Issue by ravicious** [electron/electron#40173]
https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app

The user can click the deep link either when the app is already open or when it's yet to be started – clicking the link will cause the OS to open the app.

On macOS, both scenarios will go through the same `open-url` event. Well, that's as long as you register a listener for this event fast enough. Initially I did so from an async function and that one extra tick was too late, I had to move it to sync code executed during the start of the app.

On Windows and Linux however, those two scenarios differ. If your app is already running, `second-instance` is going to be fired. But what if the app is closed? The docs don't explain that, but the way to solve it is to read `process.argv` at the start of the app and grab the URL from there.

Moreover, the docs use this approach to calling `app.setAsDefaultProtocolClient`:

```javascript
if (process.defaultApp) {
  if (process.argv.length >= 2) {
    app.setAsDefaultProtocolClient('electron-fiddle', process.execPath, [path.resolve(process.argv[1])])
  }
} else {
  app.setAsDefaultProtocolClient('electron-fiddle')
}
```

They do not explain why this particular way of calling this method is required. I suppose it's to enable protocol handling on Windows when running a "dev" version of the app that is not packaged? On macOS and Linux, custom protocol handling requires a packaged app anyway.

See other people confused about how to implement this properly:

* #33032

**Comment by felixrieseberg** [comment]
Thanks for the feedback, which is honestly fair - it *is* a little hard to get started with handling custom protocols. We could for sure improve the documentation or improve the actual API.

**Comment by shreykul** [comment]
Hi @felixrieseberg, Can I try improving the documentation?

**Comment by ravicious** [comment]
Another piece of feedback related to the deep link docs:

As I mentioned already, the docs don't show how to handle cold start from a deep link on Windows and Linux. This will likely involve checking `process.argv`. The docs use `dialog.showErrorBox` to demonstrate that the deep link handlers are working.

I ended up adding a handler for cold start before `app.whenReady` but I still used `dialog.showErrorBox` to check if it works. Unfortunately for me, this is the behavior of `dialog.showErrorBox` on Linux:

> This API can be called safely before the `ready` event the `app` module emits, it is usually used to report errors in early stage of startup. If called before the app `ready` event on Linux, the message will be emitted to stderr, and no GUI dialog will appear.

When opening a deep link, the app is not started from the terminal of course, so there's no way to check stderr in that case. So if the docs were to show how to add a handler for a cold start, they'd either have to execute it after `app.whenReady` or use something else than `dialog.showErrorBox` to show that it's working.

**Comment by Narendherraj** [comment]
Is this issue closed now ?

**Comment by ravicious** [comment]
Nope, the docs haven't been updated since the issue was made. The three main points I mentioned in the issue are yet to be addressed:

* The need to add a listener for `open-url` synchronously on macOS, within the same tick as the main body of your main.js.
* The need to read argv on Windows and Linux on a "cold" start.
* The purpose of the complicated call to `app.setAsDefaultProtocolClient`.

**Comment by Ekam219** [comment]
This can be achieved by directly adding the event listener during the initialization of your Electron app, typically at the beginning of your main.js file
const { app } = require('electron');

// Add event listener for 'open-url' synchronously
app.on('open-url', (event, url) => {
    // Handle the URL appropriately
    console.log('Received URL:', url);
});
it's crucial to read the command line arguments (argv) to extract the deep link URL. This can be done at the beginning of your main.js file before any other initialization logic. 
const { app } = require('electron');

// Read command line arguments (argv) for deep link handling
const handleDeepLinkOnColdStart = () => {
    const url = process.argv.find(arg => arg.startsWith('your-protocol://'));
    if (url) {
        // Handle the URL appropriately
        console.log('Received URL on cold start:', url);
    }
};

// Call the function to handle deep links on cold start
handleDeepLinkOnColdStart();

**Comment by zehor-l** [comment]
const { app } = require('electron');
const path = require('path');

// Event listener for deep links
app.on('open-url', (event, url) => {
  // Handle the deep link URL here
});

// Set as default protocol client
if (process.defaultApp && process.argv.length >= 2) {
  app.setAsDefaultProtocolClient('your-protocol', process.execPath, [path.resolve(process.argv[1])]);
} else {
  app.setAsDefaultProtocolClient('your-protocol');
}

**Comment by mathewalexKerala** [comment]
is this still on ? Can I work on this ?

**Comment by ravicious** [comment]
I think the issue is yet to be fixed. If it was done, someone would have updated its status.

You're free to work on this I believe. See the summary of the main three points that need to be addressed: https://github.com/electron/electron/issues/40173#issuecomment-1913095365

**Comment by shah0108** [comment]
is this still on ? Can I work on this ?

**Comment by Shivam6195** [comment]
i am interested to work on it , may i ?

**Comment by izzzabellay** [comment]
Is it still on? May I work on this as a part of my school assignment?

**Comment by jatinsingh2629** [comment]
hello, i am interested to work on this project, may I? and this issue is still there or not?

**Comment by kallal79** [comment]
I'd like to work on this issue because I specialize in reproducible infrastructure and cross-platform clarity. I’ve resolved CI and environment setup issues in Veraison and contributed to post-quantum cryptography tooling, so I’m comfortable improving documentation and debugging platform-specific startup behavior.

Plan of Action:

Update the deep link documentation to clarify:

Why open-url must be registered synchronously on macOS.

How to extract deep link URLs from process.argv on Windows/Linux cold starts.

The rationale behind conditional app.setAsDefaultProtocolClient usage for dev vs packaged apps.

Provide annotated code snippets for each platform scenario.

Replace dialog.showErrorBox with a cross-platform alternative for early-stage debugging visibility.

**Comment by Maverick-666** [comment]
Hi maintainers, I'd like to work on this issue if it's still available.

I plan to update the deep link docs and linked fiddle example to cover:
1) synchronous `open-url` listener registration on macOS,
2) `process.argv` handling for Windows/Linux cold starts,
3) clearer explanation of `app.setAsDefaultProtocolClient` for dev vs packaged apps.

I'll open a PR shortly. Thanks!

**Comment by Maverick-666** [comment]
Thanks! The commit has been re-signed and force-pushed. Could a maintainer please approve workflows for this fork PR and take a review when convenient?

**Comment by damianremington274-blip** [comment]
A quick fix could be adjusting this parameters

**Comment by Maverick-666** [comment]
Thanks for the suggestion.

Could you share the exact parameter(s) you mean (file + line), and what behavior you expect to change?

I already opened PR #50270 for this issue. If you have a concrete tweak in mind, please comment directly on that PR diff so we can evaluate it precisely.

**Comment by nikwen** [comment]
@Maverick-666 The person you replied to was a bot, unfortunately. Don't think too much about their message.

**Comment by Maverick-666** [comment]
@nikwen Thanks for clarifying — got it. I’ll ignore that bot message.

**Comment by spencerwalsh651-cmyk** [comment]
Why do u think I was a bot lol

**Comment by Maverick-666** [comment]
Thanks for clarifying — my bad. I’ll focus on maintainers’ technical feedback in this thread.

**Comment by santichausis** [comment]
I'll take a stab at the two remaining gaps here — the cold-start argv handling was already addressed in #49142, so this covers:

1. A note explaining why the `process.defaultApp` branch is needed when calling `app.setAsDefaultProtocolClient` (dev vs. packaged app on Windows).
2. A callout on why the macOS `open-url` listener must be registered synchronously at startup (Electron doesn't queue the event for a listener that isn't attached yet).

Opening a PR shortly.
